### PR TITLE
feature/SKALE-2814 partial catch up

### PR DESCRIPTION
**Pull request by sergiy-skalelabs**
Partial catch up, DB safety issues.
If **skaled** crashes after some transactions executed and their **state data** was saved into **level DB** (i.e. block was not finished) then next startup these transactions are skipped and only rest of transactions are executed to finish block mining.

**NOTICE:** the new `Contract Events Integrity Test` was used to verify partial catch-up data, you can find it here:
https://github.com/skalenetwork/SkaleExperimental/tree/master/skaled-tests/test-event-integrity

### CODE REVIEW

File: libethereum/Client.cpp [hunk 1]
```diff
@@ -226,6 +226,8 @@ void Client::init( WithExisting _forceAction, u256 _networkId ) {
     if ( m_dbPath.size() )
         Defaults::setDBPath( m_dbPath );
 
+    bc().m_stateDB = m_state.db();
+
```
Comment (dimalit) [inline libethereum/Client.cpp:239]: Why was this necessary?

File: libskale/State.cpp [hunk 2]
```diff
@@ -817,7 +824,19 @@ std::pair< ExecutionResult, TransactionReceipt > State::execute( EnvInfo const&
 
         removeEmptyAccounts = _envInfo.number() >= _sealEngine.chainParams().EIP158ForkBlock;
         commit( removeEmptyAccounts ? State::CommitBehaviour::RemoveEmptyAccounts :
-                                      State::CommitBehaviour::KeepEmptyAccounts );
+                                      State::CommitBehaviour::KeepEmptyAccounts,
+            [&]( std::shared_ptr< dev::db::DatabaseFace > /*db*/,
+                std::unique_ptr< dev::db::WriteBatchFace >& writeBatch ) {
+                if ( isSaveLastTxHash ) {
```
Comment (dimalit) [inline libskale/State.cpp:846]: Why this cannot be always true?

File: libskale/State.cpp [hunk 3]
```diff
@@ -348,7 +355,7 @@ void State::commit( CommitBehaviour _commitBehaviour ) {
             }
         }
         m_db_ptr->updateStorageUsage( totalStorageUsed_ );
-        m_db_ptr->commit();
+        m_db_ptr->commit( fn_pre_commit );
```
Comment (dimalit) [inline libskale/State.cpp:374]: Why can't we pass last transaction hash to OverlayDB in the same way as we pass totalStorageUsed_?
Do we really need this callback mechanism?

**Review by dimalit — COMMENTED** [review]
Where does this use new ConsensusEngine::getBlock() API call?
(see also other comments)

File: libethereum/Client.cpp (hunk 1 above)
Comment (sergiy-skalelabs) [reply libethereum/Client.cpp:239]: This is initialization of variable which must be non-null smart pointer earlier

File: libskale/State.cpp (hunk 2 above)
Comment (sergiy-skalelabs) [reply libskale/State.cpp:846]: I am just providing new and old versions of methods at once
isSaveLastTxHash = true -> my new version needed for partial catch up

File: libskale/State.cpp (hunk 3 above)
Comment (sergiy-skalelabs) [reply libskale/State.cpp:374]: We cannot pass last transaction hash. This is not what is needed here. We need to do operations which modify batch actions created in memory. This is same as transactions in classic SQL. Function handler to add level DB batch actions looks like the only way

File: libskale/State.cpp (hunk 2 above)
Comment (dimalit) [reply libskale/State.cpp:846]: So can we just remove this parameter and work as if it's value is true?

File: libskale/State.cpp (hunk 3 above)
Comment (dimalit) [reply libskale/State.cpp:374]: Yes, of course we need it in batch. But what's the problem to have "last transaction hash" field in OverlayDB class and have it put this field writing to DB into batch?

File: libethereum/Client.cpp (hunk 1 above)
Comment (dimalit) [reply libethereum/Client.cpp:239]: I meant this:
`bc().m_stateDB = m_state.db();`

This looks like intruding into BlockChain class' private parts..

File: libskale/State.cpp (hunk 2 above)
Comment (sergiy-skalelabs) [reply libskale/State.cpp:830]: yes, right, but I need it initialized

File: libethereum/Client.cpp (hunk 1 above)
Comment (sergiy-skalelabs) [reply libethereum/Client.cpp:239]: yes, right, but I need it initialized

File: libskale/State.cpp (hunk 2 above)
Comment (sergiy-skalelabs) [reply libskale/State.cpp:846]: State::execute() is invoked several times in code and I am sure I need new isSaveLastTxHash parameter to be true only in one place, and not sure about other places, so isSaveLastTxHash is too early to remove

**Review by dimalit — COMMENTED** [review]
1. I'm not yet sure, that in current implementation we won't get duplicated receipts.
2. Post-commit callback can be eliminated from State - will do it tomorrow
3. Still don't see where particular bloom filter can be corrupted. Will look more

**Comment by sergiy-skalelabs** [comment]
> 1. _I'm not yet sure, that in current implementation we won't get duplicated receipts._
**ANSWER:** we cannot get duplicated receipts, everything is detected at right place
> 2. _Post-commit callback can be eliminated from State - will do it tomorrow_
**ANSWER:** need to discuss this
> 3. _Still don't see where particular bloom filter can be corrupted. Will look more_
**ANSWER:** it's not corrupted, it's absent time to time in log search results, but saved into extras DB always

**Review by kladkogex — APPROVED** [review]
